reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1227 @@
+//
+// Note that the default options can result in ever-increasing memory usage
+// unless one of the Flush* methods is called on a regular basis.
+var DefaultAssemblerOptions = AssemblerOptions{
+	MaxBufferedPagesPerConnection: 0, // unlimited
+	MaxBufferedPagesTotal:         0, // unlimited
+}
+
+// AssemblerOptions controls the behavior of each assembler.  Modify the
+// options of each assembler you create to change their behavior.
+type AssemblerOptions struct {
+	// MaxBufferedPagesTotal is an upper limit on the total number of pages to
+	// buffer while waiting for out-of-order packets.  Once this limit is
+	// reached, the assembler will degrade to flushing every connection it
+	// gets a packet for.  If <= 0, this is ignored.
+	MaxBufferedPagesTotal int
+	// MaxBufferedPagesPerConnection is an upper limit on the number of pages
+	// buffered for a single connection.  Should this limit be reached for a
+	// particular connection, the smallest sequence number will be flushed, along
+	// with any contiguous data.  If <= 0, this is ignored.
+	MaxBufferedPagesPerConnection int
+}
+
+// Assembler handles reassembling TCP streams.  It is not safe for
+// concurrency... after passing a packet in via the Assemble call, the caller
+// must wait for that call to return before calling Assemble again.  Callers can
+// get around this by creating multiple assemblers that share a StreamPool.  In
+// that case, each individual stream will still be handled serially (each stream
+// has an individual mutex associated with it), however multiple assemblers can
+// assemble different connections concurrently.
+//
+// The Assembler provides (hopefully) fast TCP stream re-assembly for sniffing
+// applications written in Go.  The Assembler uses the following methods to be
+// as fast as possible, to keep packet processing speedy:
+//
+// Avoids Lock Contention
+//
+// Assemblers locks connections, but each connection has an individual lock, and
+// rarely will two Assemblers be looking at the same connection.  Assemblers
+// lock the StreamPool when looking up connections, but they use Reader
+// locks initially, and only force a write lock if they need to create a new
+// connection or close one down.  These happen much less frequently than
+// individual packet handling.
+//
+// Each assembler runs in its own goroutine, and the only state shared between
+// goroutines is through the StreamPool.  Thus all internal Assembler state
+// can be handled without any locking.
+//
+// NOTE:  If you can guarantee that packets going to a set of Assemblers will
+// contain information on different connections per Assembler (for example,
+// they're already hashed by PF_RING hashing or some other hashing mechanism),
+// then we recommend you use a seperate StreamPool per Assembler, thus
+// avoiding all lock contention.  Only when different Assemblers could receive
+// packets for the same Stream should a StreamPool be shared between them.
+//
+// Avoids Memory Copying
```

Comment:
I'm worried that this isn't actually the case, due to the use of interfaces.  I could very well be wrong, thought.